functions/go/set-image/transformer/images.go

```diff
@@ -0,0 +1,237 @@
+			Kind:       o.GetKind(),
+			Name:       o.GetName(),
+			Namespace:  o.GetNamespace(),
+		},
+		File: &fn.File{
+			Path:  o.PathAnnotation(),
+			Index: o.IndexAnnotation(),
+		},
+		Tags: nil,
+	}
+	imageTrans.Results = append(imageTrans.Results, warning)
+}
+
+func (imageTrans *ImageTransformer) setPodSpecContainers(o *fn.KubeObject) {
+	if spec := o.GetMap("spec"); spec != nil {
+		if template := spec.GetMap("template"); template != nil {
+			if podSpec := template.GetMap("spec"); podSpec != nil {
+				for _, vecObj := range podSpec.GetSlice("containers") {
+					fieldPath := FieldPath{"spec", "template", "spec", "containers"}
+					imageTrans.updateImages(vecObj, &imageTrans.Image, o, fieldPath)
+				}
+				for _, vecObj := range podSpec.GetSlice("iniContainers") {
+					fieldPath := FieldPath{"spec", "template", "spec", "iniContainers"}
+					imageTrans.updateImages(vecObj, &imageTrans.Image, o, fieldPath)
+				}
+			}
+		}
+	}
+}
+
+func (imageTrans *ImageTransformer) setPodContainers(o *fn.KubeObject) {
+	if spec := o.GetMap("spec"); spec != nil {
+		for _, vecObj := range spec.GetSlice("containers") {
+			fieldPath := FieldPath{"spec", "containers"}
+			imageTrans.updateImages(vecObj, &imageTrans.Image, o, fieldPath)
+		}
+		for _, vecObj := range spec.GetSlice("iniContainers") {
+			fieldPath := FieldPath{"spec", "iniContainers"}
+			imageTrans.updateImages(vecObj, &imageTrans.Image, o, fieldPath)
+		}
+	}
+}
+
+func getNewImageName(oldValue string, newImage *Image) string {
+	name, tag := image.Split(oldValue)
+	if newImage.NewName != "" {
+		name = newImage.NewName
+	}
+	if newImage.NewTag != "" {
+		tag = ":" + newImage.NewTag
+	}
+	if newImage.Digest != "" {
+		tag = "@" + newImage.Digest
+	}
+	newName := name + tag
+	return newName
+}
+
+func matchImage(oldName string, oldContainer string, newImage *Image) (bool, error) {
+	// name would be deprecated later, means image name
+	if newImage.Name != "" {
+		newImage.ImageName = newImage.Name
+	}
+
+	// match image name, no container name
+	if newImage.ImageName != "" && newImage.ContainerName == "" {
+		if !image.IsImageMatched(oldName, newImage.ImageName) {
+			return false, nil
+		}
+	}
+	// match container name, no image name
+	if newImage.ImageName == "" && newImage.ContainerName != "" {
+		if oldContainer != newImage.ContainerName {
+			return false, nil
+		}
+	}
+	// match both
+	if newImage.ImageName != "" && newImage.ContainerName != "" {
+		if !image.IsImageMatched(oldName, newImage.ImageName) {
+			return false, nil
+		}
+		if oldContainer != newImage.ContainerName {
+			warning := fn.Result{
+				Message:  "container name does not match, only image name matches",
+				Severity: fn.Warning,
+			}
+			return true, warning
+		}
+	}
+	return true, nil
+}
+
+// updateImages the update process for each image, if error happened, it panics
+func (imageTrans *ImageTransformer) updateImages(o *fn.SubObject, newImage *Image, parentO *fn.KubeObject, fieldPath FieldPath) {
```

Comment:
Because I am operating upon `SubObject`, and I also need information for logging, so I have its parent KubeObject and its relative path passed in. This makes me think of the error message PR in sdk, as I try to add those information in SubObject error to help debugging. The same information is needed here again, I wonder if those fields might be part of `SubObject` some day.